wgpu-rs/src/lib.rs

```diff
@@ -455,10 +456,46 @@ impl Device {
     }
 }
 
+struct BufferMapReadAsyncUserdata {
+    size: u32,
+    callback: Box<Fn(BufferMapAsyncStatus, &[u8])>
+}
+
+struct BufferMapWriteAsyncUserdata {
+    size: u32,
+    callback: Box<Fn(BufferMapAsyncStatus, &mut [u8])>
+}
+
 impl Buffer {
     pub fn set_sub_data(&self, offset: u32, data: &[u8]) {
         wgn::wgpu_buffer_set_sub_data(self.id, offset, data.len() as u32, data.as_ptr());
     }
+
+    pub fn map_read_async<'a>(&self, start: u32, size: u32, callback: Box<Fn(BufferMapAsyncStatus, &[u8])>) {
```

Comment:
for the rust wrapper, it would be more idiomatic to write our own `BufferMapAsyncStatus<T>` that would have `Success(T)` with `T` being either `&[u8]` or `&mut [u8]` depending on the function. Otherwise, it doesn't make sense to return a data slice when the mapping failed.